util/bloom/xor.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package bloom
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+
+	"github.com/algorand/xorfilter"
+)
+
+// XorBuilder is a local alias for xorfilter.Builder
+type XorBuilder = xorfilter.Builder
+
+// XorFilter is a faster more efficient alternative to a Bloom filter
+// An XorFilter object can be used as is or with optional adittional setup.
+type XorFilter struct {
+	xor     *xorfilter.Xor8
+	holding map[uint64]bool
```

Comment:
having this map isn't the optimal data structure here. I would think that an array of arrays would be a better choice. 
But don't feel obligated to change this unless you profile it and see that it becomes an issue - in particular, you want to make sure that the number of memory allocations isn't linear with the number of elements.

Reply:
It changes the test but maintains comparability between xor and bloom. The xor filter does the majority of its work in MarshalBinary() so I had to put that in its benchmark, but then it seemed only fair to compare to to the bloom filter going through the same steps. Do we keep archived history of performance on benchmarks? In this case the benchmark isn't being used to see if changing code makes it more optimized, but to compare two different implementations.

Reply:
The xor-filter build step wants a de-duplicated set of uint64. There might be a more efficient `set<uint64>` type, but this was easy and still benchmarks faster to build the filter than the old bloom filter.

Reply:
yes, and that's fine - what I'm saying is that we should leave the de-dup to `MarshalBinary` and use the "lightest" data structure here.